4_evaluate/src/interpolate_predictions.py

```diff
@@ -0,0 +1,225 @@
+    n_batches = int(np.ceil(n_obs/batch_size))
+    # Empty arrays to populate
+    sequence_indices = np.empty(n_obs, int)
+    date_indices = np.empty(n_obs, int)
+
+    # Find prediction indices for every observation, looping over batches of
+    # observations. Without batches, this could require many GBs of memory: one
+    # byte per bool (that's how numpy store bool arrays). Batches lessen the
+    # memory burden while allowing some vectorized ops for speedup
+    for i_batch in tqdm(range(n_batches)):
+        batch_start = i_batch*batch_size
+        batch_end = min((i_batch + 1)*batch_size, n_obs)
+        this_batch_size = batch_end-batch_start
+
+        # Make one huge boolean matrix with observations on rows and sequences
+        # on columns, then call nonzero to get indices (row and column) for
+        # True values in the huge matrix. 3 conditions to satisfy: obs and
+        # sequence have the same site ID, obs is after sequence starts, and obs
+        # is before sequence ends. Use broadcasting for speedup.
+        idx_obs, idx_sequence = np.nonzero(
+            (obs_sites[batch_start:batch_end, np.newaxis] == site_ids) &
+            (obs_dates_days[batch_start:batch_end, np.newaxis] >= pred_starts_days) &
+            (obs_dates_days[batch_start:batch_end, np.newaxis] < pred_ends_days)
+        )
+        # idx_obs and idx_sequence are now two arrays of same length.
+        # Their lengths are the number of matches found between obs and sequences.
+        # idx_obs indexes the obs in this batch that match sequences, and 
+        # idx_sequence indexes the sequences that match obs in this batch.
+        # Note that one obs may match multiple sequences (since they overlap)
+        # or no sequences (since some obs are before NLDAS starts).
+
+        # Check if any sequences were found for this batch
+        if idx_obs.size == 0:
+            # If not, assign dummy value of -1 to final arrays (can't use NaN
+            # in an integer array) and move on to the next batch
+            sequence_indices[batch_start:batch_end] = -1
+            date_indices[batch_start:batch_end] = -1
+        else:
+            # Since some observations can match multiple sequences, we need to
+            # pick one sequence. We use the sequence with the earlier starting
+            # date to allow more spinup (burn-in) time.
+
+            # Select the sequence with the earliest start date
+            batch_start_dates = pred_starts_days[idx_sequence]
+            # Create a 2D array with the matches along rows and the observations in
+            # this batch along columns. Values are start dates where a match
+            # corresponds to an observation in this batch, and NaN otherwise.
+            start_date_matrix = np.where(
+                idx_obs[:, np.newaxis] == np.arange(this_batch_size),
+                batch_start_dates[:, np.newaxis],
+                np.NaN
+            )
```

Comment:
It seems to me that `idx_obs` is expected to be a shape less than `this_batch_size` due to the triple condition subsetting/index identification that happened earlier (Line 112). Is that intuition right? It seems that this would raise an error if that were the case.

For example, if the `batch_size` is 5 and `idx_obs` identifies the first observation as failing to meet the earlier conditions, we would get `idx_obs = [1, 2, 3, 4]` (right?) which would then error out when compared to `np.arange(5)`

Reply:
Your intuition is right: the length of `idx_obs` is generally not `this_batch_size`. `idx_obs` can be a different shape and this will work thanks to the broadcasting that's happening. If it were `idx_obs == np.arange(this_batch_size)` then it would error. To rephrase from the [numpy docs](https://numpy.org/doc/stable/user/basics.broadcasting.html#broadcastable-arrays):
> Here the `newaxis` index operator inserts a new axis into `idx_obs`, making it a two-dimensional nx1 array. Comparing the nx1 array with `np_arange(this_batch_size)`, which has shape `(this_batch_size,)`, yields a nx`this_batch_size` array.

It reminds me of an [outer product](https://en.wikipedia.org/wiki/Outer_product) in that one vector gets transposed and the result is a matrix where every element (i, j) is a function of the ith element of one vector and the jth element of the other.

Reply:
Ah, okay. I see, and this checks out. 

![broadcastingcheck](https://user-images.githubusercontent.com/18230221/188642790-0ad8def2-4d1f-4180-9731-f9b52dd73518.PNG)

Reply:
Nice explanatory demo!